eva_submission/eload_brokering.py

```diff
@@ -27,39 +29,46 @@ def __init__(self, eload_number: int, vcf_files: list = None, metadata_file: str
-        shutil.rmtree(output_dir)
-
-        self.upload_to_bioSamples()
-        self.broker_to_ena()
-
-    def broker_to_ena(self):
-        ena_spreadsheet = os.path.join(self._get_dir('ena'), 'metadata_spreadsheet.xlsx')
-        self.update_metadata_from_config(self.eload_cfg['validation']['valid']['metadata_spreadsheet'], ena_spreadsheet)
-        submission_file, project_file, analysis_file = process_metadata_spreadsheet(ena_spreadsheet,
-                                                                                    self._get_dir('ena'), self.eload)
-        # Upload the VCF to ENA FTP
-        ena_uploader = ENAUploader(self.eload)
-        files_to_upload = [vcf_file for vcf_file in self.eload_cfg['brokering']['vcf_files']] + \
-                          [self.eload_cfg['brokering']['vcf_files'][vcf_file]['index'] for vcf_file in self.eload_cfg['brokering']['vcf_files']]
-        ena_uploader.upload_vcf_files_to_ena_ftp(files_to_upload)
-
-        # Upload XML to ENA
-        ena_uploader.upload_xml_files_to_ena(submission_file, project_file, analysis_file)
-
-        self.eload_cfg.set('brokering', 'ena', value=ena_uploader.results)
-
-    def upload_to_bioSamples(self):
+        self.prepare_brokering(force='preparation' in brokering_tasks)
+        self.upload_to_bioSamples(force='biosamples' in brokering_tasks)
+        self.broker_to_ena(force='ena' in brokering_tasks)
+
+    def prepare_brokering(self, force=False):
+        if not self.eload_cfg.query('brokering', 'vcf_files') or force:
+            output_dir = self._run_brokering_prep_workflow()
+            self._collect_brokering_prep_results(output_dir)
+            shutil.rmtree(output_dir)
+        else:
+            self.info('Preparation has already been run, Skip!')
+
+    def broker_to_ena(self, force=False):
+        if not self.eload_cfg.query('brokering', 'ena', 'PROJECT') or force:
+            ena_spreadsheet = os.path.join(self._get_dir('ena'), 'metadata_spreadsheet.xlsx')
+            self.update_metadata_from_config(self.eload_cfg['validation']['valid']['metadata_spreadsheet'], ena_spreadsheet)
+            submission_file, project_file, analysis_file = process_metadata_spreadsheet(ena_spreadsheet,
+                                                                                        self._get_dir('ena'), self.eload)
+            # Upload the VCF to ENA FTP
+            ena_uploader = ENAUploader(self.eload)
+            files_to_upload = [vcf_file for vcf_file in self.eload_cfg['brokering']['vcf_files']] + \
+                              [self.eload_cfg['brokering']['vcf_files'][vcf_file]['index'] for vcf_file in self.eload_cfg['brokering']['vcf_files']]
+            ena_uploader.upload_vcf_files_to_ena_ftp(files_to_upload)
+
+            # Upload XML to ENA
+            ena_uploader.upload_xml_files_to_ena(submission_file, project_file, analysis_file)
+            self.eload_cfg.set('brokering', 'ena', value=ena_uploader.results)
```

Comment:
you probably also want to log if ENA brokering is already done (as in the other methods)

Reply:
Good suggestion. Done